high_performance_computing/hpc_mpi/09_porting_serial_to_mpi.md

```diff
@@ -0,0 +1,515 @@
+
+We first send our newly computed leftmost value (at position `1` in our array) to the rank preceding us.
+Since we're an odd rank, we can always assume a rank preceding us exists,
+since the earliest odd rank 1 will exchange with rank 0.
+Then, we receive the rightmost boundary value from that rank.
+
+Then, if the rank following us exists, we do the same, but this time we send the rightmost value at the end of our stick section,
+and receive the corresponding value from that rank.
+
+These exchanges mean that - as an even rank - we now have effectively exchanged the states of the start and end of our slices with our respective neighbours.
+And now we need to do the same for those neighbours (the even ranks), mirroring the same communication pattern but in the opposite order of receive/send.
+From the perspective of evens, it should look like the following (highlighting the two even ranks):
+
+![Communication strategy - even ranks first receive from odd ranks, then send to them](fig/poisson_comm_strategy_2.png)
+
+```c
+  } else {
+    // Ranks with even number receive first
+
+    if (rank != 0) {
+      // Receive data from rank-1 (if I'm not the first rank)
+      MPI_Recv(&u[0], 1, MPI_FLOAT, rank-1, 1, MPI_COMM_WORLD, &mpi_status);
+      // Send data down to rank-1
+      MPI_Send(&u[1], 1, MPI_FLOAT, rank-1, 2, MPI_COMM_WORLD);
+    }
+
+    if (rank != (n_ranks-1)) {
+      // Receive data from rank+1 (if I'm not the last rank)
+      MPI_Recv(&u[points+1], 1, MPI_FLOAT, rank+1, 1, MPI_COMM_WORLD, &mpi_status);
+      // Send data up to rank+1
+      MPI_Send(&u[points], 1, MPI_FLOAT, rank+1, 2, MPI_COMM_WORLD);
+    }
+  }
+```
+
+Once complete across all ranks, every rank will then have the slice boundary data from its neighbours ready for the next iteration.
+
+### Running our Parallel Code
+
+Now we have the parallelised code in place, we can compile and run it, e.g.:
+
+```bash
+mpicc poisson_mpi.c -o poisson_mpi
+mpirun -n 2 poisson_mpi
+```
+
+```
+Final result:
+9-8-7-6-6-5-4-3-3-2-1-0-
+Run completed in 182 iterations with residue 9.60328e-06
+```
+
+Note that as it stands, the implementation assumes that `GRIDSIZE` is divisible by `n_ranks`.
+So to guarantee correct output, we should use only factors of 12 for our `n_ranks`.
+
+
+### Testing our Parallel Code
+
+We should always ensure that as our parallel version is developed, that it behaves the same as our serial version.
+This may not be possible initially, particularly as large parts of the code need converting to use MPI, but where possible, we should continue to test. 
+So we should test once we have an initial MPI version, and as our code develops, perhaps with new optimisations to improve performance, we should test then too.
+
+:::::challenge{id=an-initial-test, title="An Initial Test"}
+Test the MPI version of your code against the serial version, using 1, 2, 3, and 4 ranks with the MPI version.

```

Comment:
I feel like we should get them to actually write a test function here with asserts, this should be the solution as well?

Reply:
Yeah, agreed. I'll add that.

Reply:
Testing locally, for some reason `math.h` isn't linked by default on Ubuntu `gcc` install, so adding a callout about the `-lm` flag.

Added a bit about using asserts to check the ranks & gridsize are compatible.
We could also add a test at the end that iterates through the `resultbuf` to make sure that it matches expectations, but it seemed like that might not be easy to do in a way that would be 'realistic'? Defensive coding I think learners can carry forwards but realistically regression testing simulation outputs would probably end up comparing output files and it'd be a bit much to rewrite the outputs to make that easier.